src/engine/renderer/InternalImage.cpp

```diff
@@ -0,0 +1,209 @@
+
+Examples of scenarii:
+
+- r_picmip set 3 with some materials having imageMinDimension set to 128:
+  attempt to reduce three times the images, but for materials having that keyword,
+  stop before the image dimension would be smaller than 128;
+- r_imageMaxDimension set to 64 with some materials having imageMinDimension set to 128:
+  reduce images having dimension greater than 64 to fit 64 dimension, but for materials
+  having that keyword, reduce image to fit 128 dimension;
+- r_imageMaxDimension set to 64 with some materials having imageMinDimension set to 32:
+  reduce images having dimension greater than 64 to fit 64 dimension, but for materials
+  having that keyword, reduce image to fit 32 dimension;
+- r_imageMaxDimension set to 64 and r_imageMinDimension set to 1024 with some materials
+  having imageMinDimension set to 128:
+  reduce images having dimension greater than 64 to fit 64 dimension, but for materials
+  having that keyword, reduce image to fit 1024 dimension;
+- r_imageMaxDimension set to 64 and r_imageMinDimension set to -1 with some materials
+  having imageMinDimension set to 128:
+  reduce images having dimension greater than 64 to fit 64 dimension, but for materials
+  having that keyword, reduce image to fit their original dimension;
+- r_picmip set to 3 and r_imageMaxDimension set to 64 with some materials having
+  imageMinDimension set to 128:
+  attempt to reduce three times the images, but if dimension would still be smaller than
+  64, reduce to fit 64 dimension, unless the image has the material keyword set;
+
+Examples of use cases:
+
+- Movie producer wanting to record a movie from a demo using the highest image definition
+  possible even if the game developper configured them to be downscaled for performance
+  purpose:
+    set r_imageMaxDimension -1
+- Low budget player wanting to reduce all textures to low definition but keep configured
+  minimum definition on the textures the game developer configured to not be reduced to a
+  given size to keep the game playable:
+    set r_imageMaxDimension 64
+- Very low budger player wanting to reduce all textures to low definition in all case:
+    set r_imageMaxDimension 32
+    set r_imageMinDimension 32
+- Competitive player wanting to reduce all textures to one flat color but keep configured
+  minimum definition on the textures the game developer configured to not be reduced to a
+  given size to keep the game playable:
+    set r_imageMaxDimension 1
+- Competitive player wanting to reduce all textures to one flat color but keep high definition
+  on the textures the game developer configured to not be reduced to keep the game playable:
+    ser r_imageMaxDimension 1
+    set r_imageMinDimension -1
+ 
+===============
+*/
+int R_GetImageCustomScalingStep( const image_t *image, const imageParams_t &imageParams )
+{
+	if ( image->bits & IF_NOPICMIP )
+	{
+		return 0;
+	}
+
+	int cvarMinDimension = std::max( -1, r_imageMinDimension->integer );
```

Comment:
I think you should get rid of `r_imageMinDimension`, it is too confusing and has little practical use.

If you want to support the use case ` Very low budget player wanting to reduce all textures to low definition in all case`, you might add a boolean cvar for whether r_imageMaxDimension overrides the shader imageMaxDimension or not. (I don't think this is actually possible in your current version of the code.)

Reply:
`r_imageMinDimension` has a legit usage, look at those examples:

This is what a “low budget” player would do (except the 2K screen) with `r_imageMaxDimension 64`, all textures would be reduced to `64`, but the buildables that would be reduced to `128` because I added those minimum limits in materials (with `128` eyes are still there):

[![image max dimension](https://dl.illwieckz.net/b/daemon/features/image-max-dimension/unvanquished_2021-02-08_033526_000.jpg)](https://dl.illwieckz.net/b/daemon/features/image-max-dimension/unvanquished_2021-02-08_033526_000.jpg)

This is what a competitive player may want to do, with `r_imageMaxDimension 1`:

[![image max dimension](https://dl.illwieckz.net/b/daemon/features/image-max-dimension/unvanquished_2021-02-08_033551_000.jpg)](https://dl.illwieckz.net/b/daemon/features/image-max-dimension/unvanquished_2021-02-08_033551_000.jpg)

But the same competitive player may still want to produce good looking things (especially if he streams his games), he would then use `r_imageMaxDimension 1` with `r_imageMinDimension 512`:

[![image max dimension](https://dl.illwieckz.net/b/daemon/features/image-max-dimension/unvanquished_2021-02-08_033621_000.jpg)](https://dl.illwieckz.net/b/daemon/features/image-max-dimension/unvanquished_2021-02-08_033621_000.jpg)

Such features are common in arena FPS, like this:

[![xonotic competitive casual](https://xonotic.org/static/img/carousel_competitive_casual.jpg)](https://xonotic.org/static/img/carousel_competitive_casual.jpg)

Usually, this can be done with agressive `r_picmip` value on renderer side, and by using `noPicmip` keyword in materials that must be kept readable. But as we see, `noPicmip` would keep a 4K texture in 4K size in all case, while the image Min/Max Dimension feature (cvar and material keyword) would allow both mappers and players to target the exact value they want.

Surfaces to be used with `imageMinDimension` are typically player and weapon models, panels to be read, grids in which the fact it looks like a grid is required for gameplay, etc. Materials with meaningful glow maps may rely on it too. The `imageMinDimension` material keyword comes as a better replacement for `noPicmip` for the mapper to provide default good values, while the `r_imageMinDimension` enables the player to have control on it.

Note: the `-1` values are mostly there for debugging purposes, we probably don't need to add boolean cvars for disabling those features, and we have no need to put those values in GUI.

Reply:
OK so the semantics of a positive r_imageMinDimension are: do nothing if shader imageMinDimension is unset; overrwrite the shader value if it is set. That behavior just seems so random. Why is such a behavior needed? You already have the better version of nopicmip by using the shader parameter alone. Also this behavior means you can get a texture drawn with higher resolution than imageParams.imageMaxDimension if you set r_imageMinDimension to a positive value.

I would expect setting `imageMinDimension 1` in a shader to have the same effect as not setting it - i.e. arbitrarily down-scalable. The strange behavior of r_imageMinDimension means the effect would be different.

Reply:
> Also this behavior means you can get a texture drawn with higher resolution than imageParams.imageMaxDimension if you set r_imageMinDimension to a positive value.

I don't really see as an issue. It's expected a material keyword provides the common behavior and the cvar can override it (even to do things that do not make much sense for the general use case).

Reply:
The way I see it, the shader parameters define a window of allowable dimensions for the image. If the minimum dimension is omitted it should be equivalent to writing 1. If the maximum dimension is omitted it should be equivalent to writing an arbitrarily large number. So the use case "Competitive player wanting to reduce all textures to one flat color but keep high definition
  on the textures the game developer configured to not be reduced" is philosophically wrong and should not be supported.

In the movie producer use case, the desire is to ignore shader max dimensions. In the very low-budget player use case, the desire is to ignore shader min dimensions. So I propose to get rid of `r_imageMinDimension` (and the -1 special case of `r_imageMaxDimension`) and add instead two booleans, `r_ignoreShaderMinDimension` and `r_ignoreShaderMaxDimension`. The code could be as follows:
```

if ( ( image->bits & IF_NOPICMIP ) )
{
    return 0;
}
int minDimension = imageParams.minDimension > 0 && !r_ignoreShaderMinDimension->integer
                   ? imageParams.minDimension
                   : 1;
int maxDimension = imageParams.maxDimension > 0 && !r_ignoreShaderMaxDimension->integer
                   ? imageParams.maxDimension
                   : std::numeric_limits<int>::max();
if ( r_imageMaxDimension->integer > 0 )
{
    maxDimension = std::min( maxDimension, r_imageMaxDimension->integer );
}
// Consider the larger edge as the "image dimension"
int scaledDimension = std::max( image->width, image->height );
int scalingStep = 0;
// 1st priority: scaledDimension >= minDimension
// 2nd priority: scaledDimension <= maxDimension
// 3rd priority: scalingStep >= r_picMip->integer
// 4th priority: scalingStep as low as possible
while ( scaledDimension > maxDimension || scalingStep < r_picMip->integer )
{
    scaledDimension >>= 1;
    if ( scaledDimension < minDimension ) break;
    ++scalingStep;
}
return scalingStep;
```

Reply:
> So the use case "Competitive player wanting to reduce all textures to one flat color but keep high definition on the textures the game developer configured to not be reduced" is philosophically wrong and should not be supported.

If you believe this need (which is a serious and real need) is philosophically wrong, maybe that means the cvar is misnamed. Maybe they must be named `r_imageOverrideMinDimension` and `r_imageOverrideMaxDimension`.

Because it's perfectly legit for some non-exotic use case like an arena shooter to flat out textures but wanting at the same time to look good. The “competitive player” may want to flat out textures for two reasons: 1. performance, 2, readability. The `imageMinDimension` material keyword is provided for the mapper to ensure the texture is properly rendered for gameplay or readability purpose (like seeing grates as grates and not as glasses, distinguishes a face, reading sign posts…) so it has an implicit meaning of “this texture suffers from downscaling” and the value means the texture suffers too much from downscaling under this value.

So it's perfectly valid for a player who wants to flat out textures for visibility purpose to prevent textures known to suffer from downscaling (and marked accordingly) to not use the minimum-size known (which can be pretty ugly) but an higher size.

An image not marked with `imageMinDimension` means it may not affect readability or gameplay even when reduced to 1×1, an image marked with `imageMinDimension` means it may affect readability or gameplay if reduced under the given value, it does not mean it looks good at this value.

I first designed this for Unvanquished but because that's an engine feature I want to make sure the option is not only designed for Unvanquished special use case. For example for Unvanquished we may want to forbid the reduction of textures size for visibility purpose, but that can be done by another cvar (like `r_imageRequiredMinDimension` a game can rely on or not, with a `CVAR_CHEAT` flag, independent to the core feature of downscaling control.

What is said to be “philosophically wrong and should not be supported” is the reason why `noPicmip` exists maybe since 20 years, and this is a pretty valid need. What brings this PR is more control on those pictures known to suffer from downscaling, the user would be able to ask for a quality he considers to be high enough for good looking, even if smaller than the default size of the image.

At some point we may even want to add a cvar to set a maximum size for images flagged with nopicmip for more control, like `r_imageOverridePicmipMaxDimension`, but that would mainly be for legacy assets not relying on `imageMinDimension`. This is just to tell about the possibilities (it's needed to take this eventuality in account when designing what is currently proposed) I have not plan on it at this point but I take care of not closing that door. I can see another need for this though: some images are flagged `RSF_NOMIP` without having a materials attached to them (so `imageMinDimension` can't be used), such cvar would control them.

Also it looks OK to me if cvar can make things renders in unexpected way (we can even disable glow mapping which can have negative impact on gameplay or readability), what is important is that assets looks good by default and provided presets use safe options.

Reply:
The other competitive player use case is still supported in my version - the one where you put things with a minimum dimension to the minimum, and everything else to 1x1. `r_imageMaxDimension 1` (or `r_picmip 99` would also do the trick).

It is not the case that `noPicmip` is philosophically wrong in my interpretation. `noPicmip` is equivalent to setting the minimum and maximum dimensions of the window to the actual size of the texture.

The thing that I say is philosophically wrong is that a shader with no dimension specifications is treated differently than one with `imageMinDimension 1` and `imageMaxDimension 999999`.

Reply:
I have tested your code suggestion and I have not found a way to flat out the textures (reduce to `1×1`)  while keeping “textures that are known to suffer from downscaling” in their native resolution (or one arbitrarily high enough).
Also I was able to set `r_imageMaxDimension` to `1` at the same time I've set `r_imageMinDimension` to `1024`.

If we're going to split the cvars for easier understanding, then we also need `r_overrideShaderMinDimension` and `r_overrideShaderMaxDimension`.

Reply:
I may have an implementation that can fulfill all the needs, with 6 cvars.

- `r_imageMinDimension`
- `r_imageMaxDimension`
- `r_ignoreShaderMinDimension`
- `r_ignoreShaderMaxDimension`
- `r_overrideShaderMinDimension`
- `r_overrideShaderMaxDimension`

Note: I'm not sure `overrideShaderMinDimension` is a good name, maybe `force`, `replace` instead of `override`? I don't know.

I'll push my implementation later.

Reply:
> I have tested your code suggestion and I have not found a way to flat out the textures (reduce to `1×1`) while keeping “ètextures that are known to suffer from downscalingè” in their native resolution (or one arbitrarily high enough).

Right, you can't reduce non-dimension-spec textures to 1x1 while keeping others at max resolution in my version. You can have some at 1x1 while reducing others to configured `minDimension`.

> Also I was able to set `r_imageMaxDimension` to `1` at the same time I've set `r_imageMinDimension` to `1024` (the later was not used in any way and that was confusing).

My implementation does not have `r_imageMinDimension`.

Reply:
> My implementation does not have `r_imageMinDimension`.

Oh yes, I'm dumb.

Well, with the `r_overrideShaderMinDimension` cvar I added, I can set an arbitrary high value, but not unlock the unknown default resolution for _the one known to suffer from downscaling_ while downscaling the others. People may still set `9999` but that's not clean. That's not a big problem though, but that's not clean.

The problem was about:

- getting a fixed dimension target for downscaling (like, if target is `256`, do nothing with `128` textures, divide once `512` textures, twice `1024` textures, instead of dividing once both `4096` textures and `128`, ending with a mix of textures with `2048` and `64` dimensions,
- making possible for artists to set boundaries on select textures to ensure they are readable in game (or look good enough)
- giving fine control to the user on the engine side for generic textures
- giving fine control to the user on the engine side for textures with boundaries

At least the current implementation of the PR (I still have to edit the comments in code to reflect current code) seems to fulfill the need.